mantidproject/mantid
# Muon - Load File Widget (Single Thread)

closes #23322, removes PR #23329 

This PR creates a new "load file widget" for use in the new Muon GUI. See the issue for a description of the main features.

This PR replaces the PR #23329, by reverting back to an earlier version which uses the `thread_model.py` code (rather than the multi-threaded `threading_manager.py`, which has also been removed via the PR #23296). Functionality removed : Copy-to-clipboard button, multi-threading, progress bar for loading, cancel button for loading.

Essentially the widget allows for a user to load a single or multiple files, via a Browse button or via explicitly entering the filenames into the widget.

Code Review
===========

Several common utility functions have been added and unit tested ;

```
muon_file_utils.py
run_string_utils.py
```

A class to store the loaded data has also been implemented and tested;

```
muon_load_data.py
```

The widget itself is in an MVP pattern (and tested primarily through the presenter);

```
loadfile/load_file_view.py
loadfile/load_file_model.py
loadfile/load_file_presenter.py
```

Please give your advice regarding:

- Adequate coverage of my unit tests
- Readability of code, e.g. function names.
- Organisation and structure of code, e.g. separation of concerns.

Testing
=======

The unit tests on the presenter are a good reference point for the functionality I have explicitly coded into the widget.

The widget can be tested without Mantid using the following test code (for example add a file `test.py` into `/scripts/`)

```Python
from __future__ import (absolute_import, division, print_function)

from qtpy import QtGui, QtCore, QtWidgets
from mantid.simpleapi import *
from Muon.GUI.MuonAnalysis.loadfile.load_file_model_multithreading import BrowseFileWidgetModel
from Muon.GUI.MuonAnalysis.loadfile.load_file_view import BrowseFileWidgetView
from Muon.GUI.MuonAnalysis.loadfile.load_file_presenter import BrowseFileWidgetPresenter

from Muon.GUI.Common.muon_load_data import MuonLoadData

if __name__ == "__main__":
    import sys

    app = QtWidgets.QApplication(sys.argv)
    ui = BrowseFileWidgetPresenter(BrowseFileWidgetView(), BrowseFileWidgetModel(MuonLoadData()))
    ui.show()
    sys.exit(app.exec_())
```

Which will display the widget. Specific things to test:

- Try to cause a crash in any way possible.
- Try to load multiple files by using the browse button. Check the console output to ensure the Load algorithm is being run.
- Try to load multiple files by entering them by hand, as a ";" separated list.
- Check that the copy to clipboard icon works.
- Check that browsing for files with files already loaded **appends** the files to the list.
- Check that you can load more than 10 files, and that the behaviour from this is understandable.
- Testing all the above with multi-period data (see comments in this PR for which runs are multi-period for each supported instrument)



Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
